File: src/main/java/hiiragi283/material/api/machine/IMachineProperty.java
package hiiragi283.material.api.machine;

import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.common.util.INBTSerializable;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Consumer;

@ParametersAreNonnullByDefault
public interface IMachineProperty extends INBTSerializable<NBTTagCompound> {

    default int getProcessTime() {
        return 100;
    }

    default int getEnergyRate() {
        return 32;
    }

    default int getEnergyCapacity() {
        return getProcessTime() * getEnergyRate() * 5;
    }

    default int getItemSlotCounts() {
        return 1;
    }

    default int getFluidSlotCounts() {
        return 0;
    }

    default Set<ModuleTrait> getModuleTraits() {
        return new HashSet<>();
    }

    String KEY_PROCESS = "ProcessTime";
    String KEY_RATE = "EnergyRate";
    String KEY_ITEM = "ItemSlots";
    String KEY_FLUID = "FluidSlots";

    static IMachineProperty of() {
        return of(HiiragiUtil.getEmptyConsumer());
    }

    static IMachineProperty of(NBTTagCompound nbt) {
        return of(property -> {
            if (nbt.hasKey(KEY_PROCESS)) property.processTime = nbt.getInteger(KEY_PROCESS);
            if (nbt.hasKey(KEY_RATE)) property.energyRate = nbt.getInteger(KEY_RATE);
            if (nbt.hasKey(KEY_ITEM)) property.itemSlots = nbt.getInteger(KEY_ITEM);
            if (nbt.hasKey(KEY_FLUID)) property.fluidSlots = nbt.getInteger(KEY_FLUID);
        });
    }

    static IMachineProperty of(Consumer<IMachineProperty.Impl> consumer) {
        var property = new IMachineProperty.Impl();
        consumer.accept(property);
        return property;
    }

    //    INBTSerializable    //

    @Override
    default NBTTagCompound serializeNBT() {
        var tag = new NBTTagCompound();
        tag.setInteger(KEY_PROCESS, getProcessTime());
        tag.setInteger(KEY_RATE, getEnergyRate());
        tag.setInteger(KEY_ITEM, getItemSlotCounts());
        tag.setInteger(KEY_FLUID, getFluidSlotCounts());
        return tag;
    }

    class Impl implements IMachineProperty {

        public int processTime = 20 * 5;
        public int energyRate = 32;
        public int itemSlots = 1;
        public int fluidSlots = 0;

        @Override
        public int getProcessTime() {
            return processTime;
        }

        @Override
        public int getEnergyRate() {
            return energyRate;
        }

        @Override
        public int getItemSlotCounts() {
            return itemSlots;
        }

        @Override
        public int getFluidSlotCounts() {
            return fluidSlots;
        }

        //    INBTSerializable    //

        @Override
        public void deserializeNBT(NBTTagCompound nbt) {
            if (nbt.hasKey(KEY_PROCESS)) processTime = nbt.getInteger(KEY_PROCESS);
            if (nbt.hasKey(KEY_RATE)) energyRate = nbt.getInteger(KEY_RATE);
            if (nbt.hasKey(KEY_ITEM)) itemSlots = nbt.getInteger(KEY_ITEM);
            if (nbt.hasKey(KEY_FLUID)) fluidSlots = nbt.getInteger(KEY_FLUID);
        }

    }

}

File: src/main/kotlin/hiiragi283/material/api/machine/IMachineProperty.kt
package hiiragi283.material.api.machine

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import hiiragi283.material.util.HiiragiJsonSerializable
import hiiragi283.material.util.getIntegerOrNull
import hiiragi283.material.util.getStringOrNull
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraft.nbt.NBTTagString
import net.minecraftforge.common.util.Constants

interface IMachineProperty : HiiragiJsonSerializable {

    val recipeType: MachineType
    val processTime: Int
    val energyRate: Int
    val itemSlots: Int
    val fluidSlots: Int
    val machineTraits: Set<MachineTrait>

    fun getEnergyRequired(): Int = processTime * energyRate

    fun getEnergyCapacity(): Int = getEnergyRequired() * 5

    fun serialize() = NBTTagCompound().also { tag: NBTTagCompound ->
        tag.setString(KEY_TYPE, recipeType.name)
        tag.setInteger(KEY_PROCESS, processTime)
        tag.setInteger(KEY_RATE, energyRate)
        tag.setInteger(KEY_ITEM, itemSlots)
        tag.setInteger(KEY_FLUID, fluidSlots)
        val tagList = NBTTagList()
        machineTraits
            .map(MachineTrait::name)
            .map(::NBTTagString)
            .forEach { tagList.appendTag(it) }
        tag.setTag(KEY_TRAIT, tagList)
    }

    //    HiiragiJsonSerializable    //

    override fun getJsonElement(): JsonElement {

        val root = JsonObject()

        root.addProperty("recipe_type", recipeType.name)
        root.addProperty("process_time", processTime)
        root.addProperty("energy_rate", energyRate)
        root.addProperty("item_slots", itemSlots)
        root.addProperty("fluid_slots", fluidSlots)

        val traitsJson = JsonArray()
        machineTraits.map(MachineTrait::name).forEach(traitsJson::add)
        root.add("machine_traits", traitsJson)

        return root

    }

    companion object {

        const val KEY_TYPE = "RecipeType"
        const val KEY_PROCESS = "ProcessTime"
        const val KEY_RATE = "EnergyRate"
        const val KEY_ITEM = "ItemSlots"
        const val KEY_FLUID = "FluidSlots"
        const val KEY_TRAIT = "MachineTraits"

        fun of(tag: NBTTagCompound) = of(
            recipeType = tag.getStringOrNull(KEY_TYPE)?.let(MachineType.Companion::from) ?: MachineType.NONE,
            processTime = tag.getIntegerOrNull(KEY_PROCESS) ?: 100,
            energyRate = tag.getIntegerOrNull(KEY_RATE) ?: 32,
            itemSlots = tag.getIntegerOrNull(KEY_ITEM) ?: 1,
            fluidSlots = tag.getIntegerOrNull(KEY_FLUID) ?: 0,
            machineTraits = getModuleTraits(tag)
        )

        private fun getModuleTraits(tag: NBTTagCompound): Set<MachineTrait> {
            val tagList: NBTTagList = tag.getTagList(KEY_TRAIT, Constants.NBT.TAG_STRING)
            return (0 until tagList.tagCount())
                .map(tagList::getStringTagAt)
                .mapNotNull(MachineTrait.Companion::from)
                .toSet()
        }

        fun of(
            recipeType: MachineType = MachineType.NONE,
            processTime: Int = 100,
            energyRate: Int = 32,
            itemSlots: Int = 1,
            fluidSlots: Int = 0,
            machineTraits: Set<MachineTrait> = mutableSetOf()
        ): IMachineProperty = object : IMachineProperty {
            override var recipeType: MachineType = recipeType
            override var processTime: Int = processTime
            override var energyRate: Int = energyRate
            override var itemSlots: Int = itemSlots
            override var fluidSlots: Int = fluidSlots
            override var machineTraits: Set<MachineTrait> = machineTraits
        }

    }

}